proposals/p2240.md

```diff
@@ -0,0 +1,747 @@
+
+There are a number of pending C++ standard proposals to address these issues,
+and improve variadic templates in other ways, such as
+[P1306R1: Expansion Statements](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2019/p1306r1.pdf),
+[P1858R2: Generalized Pack Declaration and Usage](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2020/p1858r2.html),
+and
+[P2277R0: Packs Outside of Templates](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2021/p2277r0.html).
+P1858R2 has been especially influential for this proposal.
+
+Swift has very recently adopted a
+[proposal](https://github.com/apple/swift-evolution/blob/main/proposals/0393-parameter-packs.md)
+to support definition-checked generic variadics. There have been several
+attempts to add such a feature to Rust, but that work is
+[currently inactive](https://github.com/rust-lang/rfcs/issues/376#issuecomment-830034029).
+
+## Proposal
+
+A "pack expansion" is a syntactic unit beginning with `...`, which is a kind of
+compile-time loop over sequences called "packs", which are specified by
+"expansion arguments". Expansion arguments can be formed from tuples with the
+`expand` operator, or from a "variadic binding", which is marked with the `each`
+keyword at the point of declaration and the point of use.
+
+The syntax and behavior of a pack expansion depends on its context, and in some
+cases by a keyword following the `...`:
+
+-   In a tuple literal expression (such as a function call argument list), `...`
+    iteratively evaluates its operand expression, and treats the values as
+    successive elements of the tuple.
+-   `...and` and `...or` iteratively evaluate a boolean expression, combining
+    the values using `and` and `or`, and ending the loop early if the underlying
+    operator short-circuits.
+-   In a statement context, `...` iteratively executes a statement.
+-   In a tuple literal pattern (such as a function parameter list), `...`
+    iteratively matches the elements of the scrutinee tuple. In conjunction with
+    variadic bindings, this enables functions to take an arbitrary number of
+    arguments.
+
+### Examples
+
+```carbon
+// Computes the sum of its arguments, which are i64s
+fn SumInts(... each param: i64) -> i64 {
+  var sum: i64 = 0;
+  ... sum += each param;
+  return sum;
+}
+```
+
+```carbon
+// Concatenates its arguments, which are all convertible to String
+fn StrCat[... each T:! ConvertibleToString](... each param: each T) -> String {
+  var len: i64 = 0;
+  ... len += each param.Length();
+  var result: String = "";
+  result.Reserve(len);
+  ... result.Append(each param.ToString());
+  return result;
+}
+```
+
+```carbon
+// Returns the minimum of its arguments, which must all have the same type T.
+//
+// Note that this implementation is not recursive. We split the parameters into
+// first and rest in order to forbid calling `Min` with no arguments.
+fn Min[T:! Comparable & Value](first: T, ... each next: T) -> T {
```

Comment:
Perhaps this should be a follow-up, but when reading this it felt somewhat unfortunate.

Both that you needed to factor out the first element and then needed to explain why you factored it out because it is somewhat surprising.

I wonder if there is a reasonable syntactic way to represent 0-or-more vs. 1-or-more?

(This isn't a blocking comment, just didn't want to forget about it. If no one has an immediate idea, maybe just worth recording it in a possible future work section?)

Reply:
I agree that it's somewhat unfortunate, but as points in mitigation:

- The comment is speaking more to the reader of the _proposal_ (flagging a difference from C++), rather than the reader of the code in a real codebase.
- If we wrote the function signature without a separate `first`, the first line would need to be something like `var result: T = (... each arg)[0];`, which doesn't express the intent nearly so clearly.

I don't have any good ideas for a syntax (maybe somehow base it on the `+` vs `*` distinction in regexes?), but that's not the only problem: We would also need to extend the typesystem to model and propagate minimum-arity constraints on segments. That seems doable, and could be desirable for other reasons as well, but it seems substantial enough to belong in a follow-up proposal.